Download error when simultaneous downloads

## Description
<!-- Describe your issue -->

How to reproduce:

1. Appstore version: `2.5.23-SNAPSHOT`
2. Open the Appstore (for any content language)
3. Download two or more apps simultaneously

https://github.com/user-attachments/assets/05b12f65-a3a6-4156-bd3f-6dd7d29bbc34

```
2025-05-07 20:14:14.473 28671-28671 Applicatio...teReceiver ai.elimu.appstore.debug              I  onReceive itemPos: 4
                                                                                                    relativeFilePath: /lang-tha/apks/ai.elimu.herufi-1001013.apk
                                                                                                    absolute filePath: /storage/emulated/0/Android/data/ai.elimu.appstore.debug/files/lang-tha/apks/ai.elimu.herufi-1001013.apk
2025-05-07 20:14:14.473 28671-28671 Applicatio...teReceiver ai.elimu.appstore.debug              I  intent: Intent { act=android.intent.action.DOWNLOAD_COMPLETE flg=0x10 pkg=ai.elimu.appstore.debug (has extras) }
2025-05-07 20:14:14.474 28671-28671 Applicatio...teReceiver ai.elimu.appstore.debug              I  onReceive itemPos: 6
                                                                                                    relativeFilePath: /lang-tha/apks/ai.elimu.filamu-1001003.apk
                                                                                                    absolute filePath: /storage/emulated/0/Android/data/ai.elimu.appstore.debug/files/lang-tha/apks/ai.elimu.filamu-1001003.apk
2025-05-07 20:14:14.475 28671-28671 Applicatio...teReceiver ai.elimu.appstore.debug              I  intent: Intent { act=android.intent.action.DOWNLOAD_COMPLETE flg=0x10 pkg=ai.elimu.appstore.debug (has extras) }
2025-05-07 20:14:14.476 28671-28671 Applicatio...teReceiver ai.elimu.appstore.debug              I  onReceive itemPos: 7
                                                                                                    relativeFilePath: /lang-tha/apks/ai.elimu.kukariri-1001010.apk
                                                                                                    absolute filePath: /storage/emulated/0/Android/data/ai.elimu.appstore.debug/files/lang-tha/apks/ai.elimu.kukariri-1001010.apk
2025-05-07 20:14:14.476 28671-28671 Applicatio...teReceiver ai.elimu.appstore.debug              I  intent: Intent { act=android.intent.action.DOWNLOAD_COMPLETE flg=0x10 pkg=ai.elimu.appstore.debug (has extras) }
2025-05-07 20:14:14.478 28671-32395 FileHelper...D5Checksum ai.elimu.appstore.debug              E  java.io.FileNotFoundException: /storage/emulated/0/Android/data/ai.elimu.appstore.debug/files/lang-tha/apks/ai.elimu.filamu-1001003.apk: open failed: EISDIR (Is a directory)
                                                                                                    	at libcore.io.IoBridge.open(IoBridge.java:574)
                                                                                                    	at java.io.FileInputStream.<init>(FileInputStream.java:179)
                                                                                                    	at ai.elimu.appstore.util.FileHelper$calculateMD5Checksum$2.invokeSuspend(FileHelper.kt:40)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                                                                                                    Caused by: android.system.ErrnoException: open failed: EISDIR (Is a directory)
                                                                                                    	at libcore.io.IoBridge.open(IoBridge.java:564)
                                                                                                    	at java.io.FileInputStream.<init>(FileInputStream.java:179) 
                                                                                                    	at ai.elimu.appstore.util.FileHelper$calculateMD5Checksum$2.invokeSuspend(FileHelper.kt:40) 
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108) 
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115) 
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684) 
2025-05-07 20:14:14.479 28671-1201  FileHelper...D5Checksum ai.elimu.appstore.debug              E  java.io.FileNotFoundException: /storage/emulated/0/Android/data/ai.elimu.appstore.debug/files/lang-tha/apks/ai.elimu.kukariri-1001010.apk: open failed: EISDIR (Is a directory)
                                                                                                    	at libcore.io.IoBridge.open(IoBridge.java:574)
                                                                                                    	at java.io.FileInputStream.<init>(FileInputStream.java:179)
                                                                                                    	at ai.elimu.appstore.util.FileHelper$calculateMD5Checksum$2.invokeSuspend(FileHelper.kt:40)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                                                                                                    Caused by: android.system.ErrnoException: open failed: EISDIR (Is a directory)
                                                                                                    	at libcore.io.IoBridge.open(IoBridge.java:564)
                                                                                                    	at java.io.FileInputStream.<init>(FileInputStream.java:179) 
                                                                                                    	at ai.elimu.appstore.util.FileHelper$calculateMD5Checksum$2.invokeSuspend(FileHelper.kt:40) 
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108) 
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115) 
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684) 
2025-05-07 20:14:14.494 28671-32395 FileHelper...D5Checksum ai.elimu.appstore.debug              E  java.io.FileNotFoundException: /storage/emulated/0/Android/data/ai.elimu.appstore.debug/files/lang-tha/apks/ai.elimu.filamu-1001003.apk: open failed: EISDIR (Is a directory)
                                                                                                    	at libcore.io.IoBridge.open(IoBridge.java:574)
                                                                                                    	at java.io.FileInputStream.<init>(FileInputStream.java:179)
                                                                                                    	at ai.elimu.appstore.util.FileHelper$calculateMD5Checksum$2.invokeSuspend(FileHelper.kt:40)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                                                                                                    Caused by: android.system.ErrnoException: open failed: EISDIR (Is a directory)
                                                                                                    	at libcore.io.IoBridge.open(IoBridge.java:564)
                                                                                                    	at java.io.FileInputStream.<init>(FileInputStream.java:179) 
                                                                                                    	at ai.elimu.appstore.util.FileHelper$calculateMD5Checksum$2.invokeSuspend(FileHelper.kt:40) 
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108) 
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115) 
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684) 
2025-05-07 20:14:14.494 28671-32384 FileHelper...D5Checksum ai.elimu.appstore.debug              E  java.io.FileNotFoundException: /storage/emulated/0/Android/data/ai.elimu.appstore.debug/files/lang-tha/apks/ai.elimu.kukariri-1001010.apk: open failed: EISDIR (Is a directory)
                                                                                                    	at libcore.io.IoBridge.open(IoBridge.java:574)
                                                                                                    	at java.io.FileInputStream.<init>(FileInputStream.java:179)
                                                                                                    	at ai.elimu.appstore.util.FileHelper$calculateMD5Checksum$2.invokeSuspend(FileHelper.kt:40)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                                                                                                    Caused by: android.system.ErrnoException: open failed: EISDIR (Is a directory)
                                                                                                    	at libcore.io.IoBridge.open(IoBridge.java:564)
                                                                                                    	at java.io.FileInputStream.<init>(FileInputStream.java:179) 
                                                                                                    	at ai.elimu.appstore.util.FileHelper$calculateMD5Checksum$2.invokeSuspend(FileHelper.kt:40) 
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108) 
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115) 
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684) 
2025-05-07 20:14:14.513 28671-28671 ApplicationListAdapter  ai.elimu.appstore.debug              I  getItemCount
2025-05-07 20:14:14.514 28671-28671 ApplicationListAdapter  ai.elimu.appstore.debug              I  getItemCount
2025-05-07 20:14:14.515 28671-28671 ApplicationListAdapter  ai.elimu.appstore.debug              I  getItemCount
```

### Device Explorer

![Device_Explorer](https://github.com/user-attachments/assets/fdf6e085-98f3-4d27-a47a-c1b637eb1929)

## Why?
<!-- Explain _why_ this issue is needed -->
If the UI allows users to download multiple APKs at once, doing so shouldn't cause an error.

---

> elimu.ai's mission is to build innovative learning software that empowers out-of-school children to teach themselves basic reading📖, writing✍🏽 and math🔢 **within 6 months**.

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListActivity.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListActivity.kt
@@ -14,6 +14,7 @@ import timber.log.Timber
 class ApplicationListActivity : AppCompatActivity() {
 
     private val TAG = "ApplicationListActivity"
+    private lateinit var appListAdapter: ApplicationListAdapter
 
     override fun onCreate(savedInstanceState: Bundle?) {
         Timber.i("onCreate")
@@ -34,8 +35,8 @@ class ApplicationListActivity : AppCompatActivity() {
 
         // Configure list adapter
         val recyclerView = findViewById<RecyclerView>(R.id.recycler_view)
-        val applicationListAdapter = ApplicationListAdapter(this)
-        recyclerView.adapter = applicationListAdapter
+        appListAdapter = ApplicationListAdapter(this)
+        recyclerView.adapter = appListAdapter
         val linearLayoutManager = LinearLayoutManager(this)
         recyclerView.layoutManager = linearLayoutManager
         val dividerItemDecoration =
@@ -50,13 +51,18 @@ class ApplicationListActivity : AppCompatActivity() {
             val applications =
                 applicationDao.loadAll()
             Timber.tag(TAG).d("applications.size(): %s", applications.size)
-            applicationListAdapter.setApplications(applications)
+            appListAdapter.setApplications(applications)
 
             val applicationVersions =
                 applicationVersionDao.loadAll()
             Timber.tag(TAG).d("applicationVersions.size(): %s", applicationVersions.size)
-            applicationListAdapter.setApplicationVersions(applicationVersions)
-            recyclerView.post(applicationListAdapter::notifyDataSetChanged)
+            appListAdapter.setApplicationVersions(applicationVersions)
+            recyclerView.post(appListAdapter::notifyDataSetChanged)
         }
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        appListAdapter.unregisterReceiver(this)
     }
 }

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListActivity.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListActivity.kt
@@ -80,7 +80,11 @@ class ApplicationListActivity : AppCompatActivity() {
 
     override fun onPause() {
         super.onPause()
-        unregisterReceiver(downloadReceiver)
+        try {
+            unregisterReceiver(downloadReceiver)
+        } catch (e: IllegalArgumentException) {
+            Timber.e(e)
+        }
     }
 
     override fun onDestroy() {

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListActivity.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListActivity.kt
@@ -2,6 +2,9 @@ package ai.elimu.appstore.ui.applications
 
 import ai.elimu.appstore.R
 import ai.elimu.appstore.room.RoomDb
+import android.app.DownloadManager
+import android.content.IntentFilter
+import android.os.Build
 import android.os.Bundle
 import androidx.appcompat.app.AppCompatActivity
 import androidx.appcompat.widget.Toolbar
@@ -15,6 +18,7 @@ class ApplicationListActivity : AppCompatActivity() {
 
     private val TAG = "ApplicationListActivity"
     private lateinit var appListAdapter: ApplicationListAdapter
+    private val downloadReceiver: DownloadCompleteReceiver = DownloadCompleteReceiver()
 
     override fun onCreate(savedInstanceState: Bundle?) {
         Timber.i("onCreate")
@@ -35,7 +39,7 @@ class ApplicationListActivity : AppCompatActivity() {
 
         // Configure list adapter
         val recyclerView = findViewById<RecyclerView>(R.id.recycler_view)
-        appListAdapter = ApplicationListAdapter(this)
+        appListAdapter = ApplicationListAdapter(this, downloadReceiver)
         recyclerView.adapter = appListAdapter
         val linearLayoutManager = LinearLayoutManager(this)
         recyclerView.layoutManager = linearLayoutManager
@@ -59,6 +63,24 @@ class ApplicationListActivity : AppCompatActivity() {
             appListAdapter.setApplicationVersions(applicationVersions)
             recyclerView.post(appListAdapter::notifyDataSetChanged)
         }
+    }
+
+    override fun onResume() {
+        super.onResume()
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            registerReceiver(downloadReceiver,
+                IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), RECEIVER_EXPORTED)
+        } else {
+            @Suppress("UnspecifiedRegisterReceiverFlag")
+            registerReceiver(downloadReceiver,
+                IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE))
+        }
+
+    }
+
+    override fun onPause() {
+        super.onPause()
+        unregisterReceiver(downloadReceiver)
     }
 
     override fun onDestroy() {

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListActivity.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListActivity.kt
@@ -86,5 +86,6 @@ class ApplicationListActivity : AppCompatActivity() {
     override fun onDestroy() {
         super.onDestroy()
         appListAdapter.unregisterReceiver(this)
+        downloadReceiver.clearListeners()
     }
 }

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
@@ -62,7 +62,7 @@ class ApplicationListAdapter(private val context: Context) :
 
             // Populate TextViews with Application details
             val application = applications!![position]
-            Timber.i("application.getPackageName(): \"" + application.packageName + "\"")
+            Timber.i("application.getPackageName(): ${application.packageName}")
             viewHolder.textViewFirstLine.text = application.packageName
             viewHolder.textViewSecondLine.text = application.applicationStatus.toString()
 
@@ -182,7 +182,7 @@ class ApplicationListAdapter(private val context: Context) :
                         viewHolder.launchButton.visibility = View.VISIBLE
                         viewHolder.launchButton.setOnClickListener {
                             Timber.i("onClick")
-                            Timber.i("Launching \"" + application.packageName + "\"")
+                            Timber.i("Launching ${application.packageName}")
                             val packageManager = context.packageManager
                             val launchIntent =
                                 packageManager.getLaunchIntentForPackage(application.packageName)

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
@@ -156,7 +156,6 @@ class ApplicationListAdapter(
                                         context
                                     )!!
                                         .isoCode + File.separator + "apks" + File.separator + apkFile.name
-                                registeredReceivers.add(downloadReceiver)
                                 request.setDestinationInExternalFilesDir(
                                     context,
                                     null,
@@ -256,8 +255,6 @@ class ApplicationListAdapter(
                                 )!!
                                     .isoCode + File.separator + "apks" + File.separator + apkFile.name
                             Timber.i("destinationInExternalFilesDir: $destinationInExternalFilesDir")
-
-                            registeredReceivers.add(downloadReceiver)
 
                             request.setDestinationInExternalFilesDir(
                                 context,

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
@@ -32,8 +32,12 @@ import com.google.android.material.snackbar.Snackbar
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
 
-class ApplicationListAdapter(private val context: Context) :
+class ApplicationListAdapter(
+    private val context: Context,
+    private val downloadReceiver: DownloadCompleteReceiver
+) :
     RecyclerView.Adapter<ApplicationViewHolder>() {
     private val layoutInflater: LayoutInflater = LayoutInflater.from(context)
 
@@ -41,6 +45,9 @@ class ApplicationListAdapter(private val context: Context) :
 
     private var applicationVersions: List<ApplicationVersion>? = null
     private val registeredReceivers: MutableSet<BroadcastReceiver> by lazy { mutableSetOf() }
+
+    private val downloadManager =
+        context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ApplicationViewHolder {
         Timber.i("onCreateViewHolder")
@@ -94,7 +101,10 @@ class ApplicationListAdapter(private val context: Context) :
                         application,
                         applicationVersions!!
                     )
-                    Timber.i("applicationVersion.getVersionCode(): %s", applicationVersion!!.versionCode)
+                    Timber.i(
+                        "applicationVersion.getVersionCode(): %s",
+                        applicationVersion!!.versionCode
+                    )
                     if (versionCodeInstalled < applicationVersion.versionCode) {
                         // An update is available for download
 
@@ -139,8 +149,6 @@ class ApplicationListAdapter(private val context: Context) :
                                 Timber.i("viewHolder.downloadUpdateButton onClick")
                                 // Initiate download of the APK file
                                 Timber.i("finalApplicationVersion.fileUrl: ${applicationVersion.fileUrl}")
-                                val downloadManager =
-                                    context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
                                 val request =
                                     DownloadManager.Request(applicationVersion.fileUrl.toUri())
                                 val destinationInExternalFilesDir =
@@ -148,24 +156,22 @@ class ApplicationListAdapter(private val context: Context) :
                                         context
                                     )!!
                                         .isoCode + File.separator + "apks" + File.separator + apkFile.name
-                                val downloadReceiver = DownloadCompleteReceiver(
-                                    position, destinationInExternalFilesDir,
-                                    checkSum = applicationVersion.checksumMd5
-                                )
-                                context.registerReceiver(
-                                    downloadReceiver,
-                                    IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE),
-                                    Context.RECEIVER_EXPORTED
-                                )
                                 registeredReceivers.add(downloadReceiver)
-                                Timber.i("destinationInExternalFilesDir: $destinationInExternalFilesDir")
                                 request.setDestinationInExternalFilesDir(
                                     context,
                                     null,
                                     destinationInExternalFilesDir
                                 )
                                 val downloadId = downloadManager.enqueue(request)
-                                Timber.i("downloadId: $downloadId")
+                                if (downloadId != -1L) {
+                                    downloadReceiver.addDownloadListener(downloadId) {
+                                        handleDownloadComplete(itemPosition = position,
+                                            relativeFilePath = destinationInExternalFilesDir,
+                                            checkSum = applicationVersion.checksumMd5
+                                        )
+                                    }
+                                    Timber.i("downloadId: $downloadId")
+                                }
 
                                 // Replace download button with progress bar
                                 viewHolder.downloadUpdateButton.visibility =
@@ -202,7 +208,10 @@ class ApplicationListAdapter(private val context: Context) :
                     if (applicationVersion == null) {
                         return
                     }
-                    Timber.i("applicationVersion.getVersionCode(): %s", applicationVersion.versionCode)
+                    Timber.i(
+                        "applicationVersion.getVersionCode(): %s",
+                        applicationVersion.versionCode
+                    )
 
                     // If the APK has been downloaded (but not yet installed), display the "Install" button
                     val apkFile =
@@ -239,8 +248,6 @@ class ApplicationListAdapter(private val context: Context) :
                             Timber.i("viewHolder.downloadButton onClick")
                             // Initiate download of the APK file
                             Timber.i("finalApplicationVersion.fileUrl: ${finalApplicationVersion.fileUrl}")
-                            val downloadManager =
-                                context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
                             val request =
                                 DownloadManager.Request(finalApplicationVersion.fileUrl.toUri())
                             val destinationInExternalFilesDir =
@@ -250,15 +257,6 @@ class ApplicationListAdapter(private val context: Context) :
                                     .isoCode + File.separator + "apks" + File.separator + apkFile.name
                             Timber.i("destinationInExternalFilesDir: $destinationInExternalFilesDir")
 
-                            val downloadReceiver = DownloadCompleteReceiver(
-                                position, destinationInExternalFilesDir,
-                                finalApplicationVersion.checksumMd5
-                            )
-                            context.registerReceiver(
-                                downloadReceiver,
-                                IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE),
-                                Context.RECEIVER_EXPORTED
-                            )
                             registeredReceivers.add(downloadReceiver)
 
                             request.setDestinationInExternalFilesDir(
@@ -267,6 +265,13 @@ class ApplicationListAdapter(private val context: Context) :
                                 destinationInExternalFilesDir
                             )
                             val downloadId = downloadManager.enqueue(request)
+                            downloadReceiver.addDownloadListener(downloadId) {
+                                handleDownloadComplete(itemPosition = position,
+                                    relativeFilePath = destinationInExternalFilesDir,
+                                    checkSum = applicationVersion.checksumMd5
+                                )
+                            }
+
                             Timber.i("downloadId: $downloadId")
 
                             // Replace download button with progress bar
@@ -291,6 +296,30 @@ class ApplicationListAdapter(private val context: Context) :
             }
         }
         return applicationVersion
+    }
+
+    private fun handleDownloadComplete(
+        itemPosition: Int,
+        relativeFilePath: String,
+        checkSum: String
+    ) {
+        CoroutineScope(Dispatchers.IO).launch {
+            val downloadedFile = context.getExternalFilesDir(relativeFilePath)
+            val downloadedFileCheckSum =
+                FileHelper.calculateMD5Checksum(downloadedFile?.absolutePath ?: "")
+            withContext(Dispatchers.Main) {
+                if (downloadedFileCheckSum != checkSum) {
+                    downloadedFile?.delete()
+                    val activity = context as? ApplicationListActivity ?: return@withContext
+                    Snackbar.make(
+                        activity.window.decorView,
+                        context.getString(R.string.download_error),
+                        Snackbar.LENGTH_LONG
+                    ).setBackgroundTint(ContextCompat.getColor(context, R.color.red)).show()
+                }
+                notifyItemChanged(itemPosition)
+            }
+        }
     }
 
     override fun getItemCount(): Int {
@@ -342,36 +371,6 @@ class ApplicationListAdapter(private val context: Context) :
             downloadProgressBar = itemView.findViewById(R.id.list_item_download_progressbar)
         }
     }
-
-
-    private inner class DownloadCompleteReceiver(private val itemPosition: Int,
-                                                 private val relativeFilePath: String,
-                                                 private val checkSum: String) :
-        BroadcastReceiver() {
-        override fun onReceive(context: Context, intent: Intent) {
-            val downloadedFile = context.getExternalFilesDir(relativeFilePath)
-            Timber.i("onReceive itemPos: $itemPosition\nrelativeFilePath: $relativeFilePath\nabsolute filePath: $downloadedFile")
-            Timber.i("intent: $intent")
-
-            CoroutineScope(Dispatchers.Main).launch {
-                val downloadedFileCheckSum = FileHelper.calculateMD5Checksum(downloadedFile?.absolutePath ?: "")
-                if (downloadedFileCheckSum != checkSum) {
-                    downloadedFile?.delete()
-                    val activity = context as? ApplicationListActivity ?: return@launch
-                    Snackbar.make(activity.window.decorView,
-                        context.getString(R.string.download_error),
-                        Snackbar.LENGTH_LONG
-                    ).setBackgroundTint(ContextCompat.getColor(context, R.color.red)).show()
-                }
-                notifyItemChanged(itemPosition)
-                this@DownloadCompleteReceiver.let { receiver ->
-                    context.unregisterReceiver(receiver)
-                    registeredReceivers.remove(receiver)
-                }
-            }
-        }
-    }
-
 
     private inner class PackageAddedReceiver(private val itemPosition: Int) : BroadcastReceiver() {
         override fun onReceive(context: Context, intent: Intent) {

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
@@ -40,6 +40,7 @@ class ApplicationListAdapter(private val context: Context) :
     private var applications: List<Application>? = null
 
     private var applicationVersions: List<ApplicationVersion>? = null
+    private val registeredReceivers: MutableSet<BroadcastReceiver> by lazy { mutableSetOf() }
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ApplicationViewHolder {
         Timber.i("onCreateViewHolder")
@@ -113,10 +114,12 @@ class ApplicationListAdapter(private val context: Context) :
                                 // Initiate installation of the APK file
                                 val intentFilter = IntentFilter(Intent.ACTION_PACKAGE_ADDED)
                                 intentFilter.addDataScheme("package")
+                                val packageAddedReceiver = PackageAddedReceiver(position)
                                 context.registerReceiver(
-                                    PackageAddedReceiver(position),
+                                    packageAddedReceiver,
                                     intentFilter
                                 )
+                                registeredReceivers.add(packageAddedReceiver)
                                 val apkUri = FileProvider.getUriForFile(
                                     context, BuildConfig.APPLICATION_ID + ".apk.provider",
                                     apkFile
@@ -145,14 +148,16 @@ class ApplicationListAdapter(private val context: Context) :
                                         context
                                     )!!
                                         .isoCode + File.separator + "apks" + File.separator + apkFile.name
+                                val downloadReceiver = DownloadCompleteReceiver(
+                                    position, destinationInExternalFilesDir,
+                                    checkSum = applicationVersion.checksumMd5
+                                )
                                 context.registerReceiver(
-                                    DownloadCompleteReceiver(
-                                        position, destinationInExternalFilesDir,
-                                        checkSum = applicationVersion.checksumMd5
-                                    ),
+                                    downloadReceiver,
                                     IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE),
                                     Context.RECEIVER_EXPORTED
                                 )
+                                registeredReceivers.add(downloadReceiver)
                                 Timber.i("destinationInExternalFilesDir: $destinationInExternalFilesDir")
                                 request.setDestinationInExternalFilesDir(
                                     context,
@@ -211,11 +216,13 @@ class ApplicationListAdapter(private val context: Context) :
                             // Initiate installation of the APK file
                             val intentFilter = IntentFilter(Intent.ACTION_PACKAGE_ADDED)
                             intentFilter.addDataScheme("package")
-                            context.registerReceiver(PackageAddedReceiver(position), intentFilter)
+                            val packageAddedReceiver = PackageAddedReceiver(position)
+                            context.registerReceiver(packageAddedReceiver, intentFilter)
                             val apkUri = FileProvider.getUriForFile(
                                 context, BuildConfig.APPLICATION_ID + ".apk.provider",
                                 apkFile
                             )
+                            registeredReceivers.add(packageAddedReceiver)
                             val intent = Intent(Intent.ACTION_INSTALL_PACKAGE)
                             intent.setData(apkUri)
                             intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
@@ -243,11 +250,16 @@ class ApplicationListAdapter(private val context: Context) :
                                     .isoCode + File.separator + "apks" + File.separator + apkFile.name
                             Timber.i("destinationInExternalFilesDir: $destinationInExternalFilesDir")
 
+                            val downloadReceiver = DownloadCompleteReceiver(
+                                position, destinationInExternalFilesDir,
+                                finalApplicationVersion.checksumMd5
+                            )
                             context.registerReceiver(
-                                DownloadCompleteReceiver(position, destinationInExternalFilesDir, finalApplicationVersion.checksumMd5),
+                                downloadReceiver,
                                 IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE),
                                 Context.RECEIVER_EXPORTED
                             )
+                            registeredReceivers.add(downloadReceiver)
 
                             request.setDestinationInExternalFilesDir(
                                 context,
@@ -298,6 +310,11 @@ class ApplicationListAdapter(private val context: Context) :
         this.applicationVersions = applicationVersions
     }
 
+    fun unregisterReceiver(context: Context) {
+        for (receiver in registeredReceivers) {
+            context.unregisterReceiver(receiver)
+        }
+    }
 
     inner class ApplicationViewHolder internal constructor(itemView: View) :
         RecyclerView.ViewHolder(itemView) {
@@ -347,6 +364,10 @@ class ApplicationListAdapter(private val context: Context) :
                     ).setBackgroundTint(ContextCompat.getColor(context, R.color.red)).show()
                 }
                 notifyItemChanged(itemPosition)
+                this@DownloadCompleteReceiver.let { receiver ->
+                    context.unregisterReceiver(receiver)
+                    registeredReceivers.remove(receiver)
+                }
             }
         }
     }
@@ -359,6 +380,8 @@ class ApplicationListAdapter(private val context: Context) :
             Timber.i("intent: $intent")
             Timber.i("intent.getData(): %s", intent.data)
             notifyItemChanged(itemPosition)
+            context.unregisterReceiver(this)
+            registeredReceivers.remove(this)
         }
     }
 }

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
@@ -170,6 +170,9 @@ class ApplicationListAdapter(
                                         )
                                     }
                                     Timber.i("downloadId: $downloadId")
+                                } else {
+                                    Timber.e("Failed to enqueue download")
+                                    notifyItemChanged(position)
                                 }
 
                                 // Replace download button with progress bar
@@ -262,11 +265,16 @@ class ApplicationListAdapter(
                                 destinationInExternalFilesDir
                             )
                             val downloadId = downloadManager.enqueue(request)
-                            downloadReceiver.addDownloadListener(downloadId) {
-                                handleDownloadComplete(itemPosition = position,
-                                    relativeFilePath = destinationInExternalFilesDir,
-                                    checkSum = applicationVersion.checksumMd5
-                                )
+                            if (downloadId != -1L) {
+                                downloadReceiver.addDownloadListener(downloadId) {
+                                    handleDownloadComplete(itemPosition = position,
+                                        relativeFilePath = destinationInExternalFilesDir,
+                                        checkSum = applicationVersion.checksumMd5
+                                    )
+                                }
+                            } else {
+                                Timber.e("Failed to enqueue download")
+                                notifyItemChanged(position)
                             }
 
                             Timber.i("downloadId: $downloadId")

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
@@ -338,7 +338,11 @@ class ApplicationListAdapter(
 
     fun unregisterReceiver(context: Context) {
         for (receiver in registeredReceivers) {
-            context.unregisterReceiver(receiver)
+            try {
+                context.unregisterReceiver(receiver)
+            } catch (e: IllegalArgumentException) {
+                Timber.e(e)
+            }
         }
     }
 

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/DownloadCompleteReceiver.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/DownloadCompleteReceiver.kt
@@ -17,6 +17,10 @@ class DownloadCompleteReceiver() : BroadcastReceiver() {
 
     override fun onReceive(context: Context, intent: Intent) {
         val downloadId = intent.getLongExtra(DownloadManager.EXTRA_DOWNLOAD_ID, -1L)
-        downloadListeners[downloadId]?.invoke()
+        downloadListeners.remove(downloadId)?.invoke()
+    }
+
+    fun clearListeners() {
+        downloadListeners.clear()
     }
 }

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/DownloadCompleteReceiver.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/DownloadCompleteReceiver.kt
@@ -26,11 +26,9 @@ class DownloadCompleteReceiver() : BroadcastReceiver() {
             val statusIndex = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS)
             if (statusIndex != -1) {
                 val status = cursor.getInt(statusIndex)
-                if (status == DownloadManager.STATUS_SUCCESSFUL) {
+                if (status == DownloadManager.STATUS_SUCCESSFUL
+                    || status == DownloadManager.STATUS_FAILED) {
                     downloadListeners.remove(downloadId)?.invoke()
-                } else {
-                    // Handle failed download if needed
-                    downloadListeners.remove(downloadId)
                 }
             }
         }

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/DownloadCompleteReceiver.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/DownloadCompleteReceiver.kt
@@ -1,0 +1,22 @@
+package ai.elimu.appstore.ui.applications
+
+import android.app.DownloadManager
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+
+typealias OnDownloadComplete = () -> Unit
+
+class DownloadCompleteReceiver() : BroadcastReceiver() {
+
+    private val downloadListeners: MutableMap<Long, OnDownloadComplete> by lazy { mutableMapOf() }
+
+    fun addDownloadListener(downloadId: Long, onComplete: OnDownloadComplete) {
+        downloadListeners.put(downloadId, onComplete)
+    }
+
+    override fun onReceive(context: Context, intent: Intent) {
+        val downloadId = intent.getLongExtra(DownloadManager.EXTRA_DOWNLOAD_ID, -1L)
+        downloadListeners[downloadId]?.invoke()
+    }
+}

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/DownloadCompleteReceiver.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/DownloadCompleteReceiver.kt
@@ -17,7 +17,24 @@ class DownloadCompleteReceiver() : BroadcastReceiver() {
 
     override fun onReceive(context: Context, intent: Intent) {
         val downloadId = intent.getLongExtra(DownloadManager.EXTRA_DOWNLOAD_ID, -1L)
-        downloadListeners.remove(downloadId)?.invoke()
+
+        val query = DownloadManager.Query().setFilterById(downloadId)
+        val downloadManager = context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
+        val cursor = downloadManager.query(query)
+
+        if (cursor.moveToFirst()) {
+            val statusIndex = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS)
+            if (statusIndex != -1) {
+                val status = cursor.getInt(statusIndex)
+                if (status == DownloadManager.STATUS_SUCCESSFUL) {
+                    downloadListeners.remove(downloadId)?.invoke()
+                } else {
+                    // Handle failed download if needed
+                    downloadListeners.remove(downloadId)
+                }
+            }
+        }
+        cursor.close()
     }
 
     fun clearListeners() {

--- a/app/src/main/java/ai/elimu/appstore/ui/applications/InitialSyncActivity.kt
+++ b/app/src/main/java/ai/elimu/appstore/ui/applications/InitialSyncActivity.kt
@@ -63,7 +63,7 @@ class InitialSyncActivity : AppCompatActivity() {
             override fun onFailure(call: Call<List<ApplicationGson>>, t: Throwable) {
                 Timber.e(t, "onFailure")
 
-                Timber.e(t, "t.getCause(): " + t.cause)
+                Timber.e(t, "t.getCause(): ${t.cause}")
 
                 // Handle error
                 Snackbar.make(binding.initialSyncTextview, t.cause.toString(), Snackbar.LENGTH_LONG).show()
@@ -91,7 +91,7 @@ class InitialSyncActivity : AppCompatActivity() {
                     // Store the new Application in the database
                     GsonToRoomConverter.getApplication(applicationGson)?.let { app ->
                         applicationDao.insert(app)
-                        Timber.i("%s%s", "Stored Application \"" + app.packageName + "\" in database with ID ", app.id)
+                        Timber.i("Stored Application ${app.packageName} in database with ID ${app.id}")
                     }
 
                     if (applicationGson.applicationStatus == ApplicationStatus.ACTIVE) {
@@ -104,10 +104,7 @@ class InitialSyncActivity : AppCompatActivity() {
                                 applicationVersionGson
                             )?.let { applicationVersion ->
                                 applicationVersionDao.insert(applicationVersion)
-                                Timber.i("%s%s", "Stored ApplicationVersion " +
-                                        applicationVersion.versionCode + " in database with ID ",
-                                    applicationVersion.id
-                                )
+                                Timber.i("Stored ApplicationVersion ${applicationVersion.versionCode} in database with ID ${applicationVersion.id}")
                             }
                         }
                     }
@@ -115,7 +112,7 @@ class InitialSyncActivity : AppCompatActivity() {
                     // Update the existing Application in the database
                     GsonToRoomConverter.getApplication(applicationGson)?.let { app ->
                         applicationDao.update(app)
-                        Timber.i("%s%s", "Updated Application \"" + app.packageName + "\" in database with ID ", app.id)
+                        Timber.i("Updated Application ${app.packageName} in database with ID ${app.id}")
                     }
 
                     // Delete all the Application's ApplicationVersions (in case deletions have been made on the server-side)
@@ -131,10 +128,8 @@ class InitialSyncActivity : AppCompatActivity() {
                                 applicationVersionGson
                             )?.let { applicationVersion ->
                                 applicationVersionDao.insert(applicationVersion)
-                                Timber.i("%s%s", "Stored ApplicationVersion " +
-                                        applicationVersion.versionCode + " in database with ID ",
-                                    applicationVersion.id
-                                )
+                                Timber.i("Stored ApplicationVersion ${applicationVersion.versionCode} " +
+                                        "in database with ID ${applicationVersion.id}")
                             }
                         }
                     }

